### ISSUE TITLE
Go 1.13 Causing "500 Internal Server Error"

### ISSUE THREAD

**Issue by marct83** [gomods/athens#1372]
**Describe the bug**
Since updating to Go 1.13 I am seeing a "500 Internal Server Error" when fetching some packages. I rolled back to go 1.12 and the issue goes away. I tested this against the public experimental version of Athens to be sure it wasn't an issue with our instance.

**Error Message**
If applicable add error message to help explain your problem or console output enclose in triple back-ticks eg,

```
export GOPROXY="https://athens.azurefd.net" && go get golang.org/x/tools/gopls@latest
go: finding golang.org/x/tools latest
go get golang.org/x/tools/gopls@latest: module golang.org/x: reading https://athens.azurefd.net/golang.org/x/@v/list: 500 Internal Server Error
```

**To Reproduce**
```export GOPROXY="https://athens.azurefd.net" && go get golang.org/x/tools/gopls@latest```

**Expected behavior**
```
get golang.org/x/tools/gopls@latest
go: finding golang.org/x/tools latest
go: downloading golang.org/x/tools v0.0.0-20190903163617-be0da057c5e3
go: extracting golang.org/x/tools v0.0.0-20190903163617-be0da057c5e3
go: downloading golang.org/x/tools v0.0.0-20190710153321-831012c29e42
go: extracting golang.org/x/tools v0.0.0-20190710153321-831012c29e42
go: downloading golang.org/x/sync v0.0.0-20190423024810-112230192c58
go: extracting golang.org/x/sync v0.0.0-20190423024810-112230192c58
go: finding golang.org/x/tools v0.0.0-20190710153321-831012c29e42
go: finding golang.org/x/sync v0.0.0-20190423024810-112230192c58
```
**Environment (please complete the following information):**
 - OS: Linux 64bit
 - Go version : 1.13
 - Proxy version : Public Experimental Version (https://athens.azurefd.net) 0.5.0(?)
 - Storage (fs/mongodb/s3 etc.) : ?

**Additional context**

**Comment by atomi** [comment]
Here are logs when trying to `go get -u github.com/gorilla/mux`

```log
time="2019-09-07T00:09:01Z" level=info msg="Exporter not specified. Traces won't be exported"
2019-09-07 00:09:01.167220 I | Starting application at port :3000
time="2019-09-07T00:09:20Z" level=error msg="exit status 1: go list -m github.com/gorilla: invalid github.com/ import path \"github.com/gorilla\"\n" http-method=GET http-path=/github.com/gorilla/@v/list http-url=/github.com/gorilla/@v/list kind="Internal Server Error" module= operation=download.ListHandler ops="[download.ListHandler pool.List protocol.List vcsLister.List]" version=
time="2019-09-07T00:09:20Z" level=error msg="exit status 1: go list -m github.com: unrecognized import path \"github.com\" (parse https://github.com?go-get=1: no go-import meta tags ())\n" http-method=GET http-path=/github.com/@v/list http-url=/github.com/@v/list kind="Internal Server Error" module= operation=download.ListHandler ops="[download.ListHandler pool.List protocol.List vcsLister.List]" version=
time="2019-09-07T00:10:07Z" level=error msg="exit status 1: go list -m github.com/gorilla: invalid github.com/ import path \"github.com/gorilla\"\n" http-method=GET http-path=/github.com/gorilla/@v/list http-url=/github.com/gorilla/@v/list kind="Internal Server Error" module= operation=download.ListHandler ops="[download.ListHandler pool.List protocol.List vcsLister.List]" version=
time="2019-09-07T00:10:08Z" level=error msg="exit status 1: go list -m github.com: unrecognized import path \"github.com\" (parse https://github.com?go-get=1: no go-import meta tags ())\n" http-method=GET http-path=/github.com/@v/list http-url=/github.com/@v/list kind="Internal Server Error" module= operation=download.ListHandler ops="[download.ListHandler pool.List protocol.List vcsLister.List]" version=
time="2019-09-07T00:10:46Z" level=error msg="exit status 1: go list -m github.com/gorilla: invalid github.com/ import path \"github.com/gorilla\"\n" http-method=GET http-path=/github.com/gorilla/@v/list http-url=/github.com/gorilla/@v/list kind="Internal Server Error" module= operation=download.ListHandler ops="[download.ListHandler pool.List protocol.List vcsLister.List]" version=
time="2019-09-07T00:21:57Z" level=error msg="exit status 1: go list -m github.com/gorilla: invalid github.com/ import path \"github.com/gorilla\"\n" http-method=GET http-path=/github.com/gorilla/@v/list http-url=/github.com/gorilla/@v/list kind="Internal Server Error" module= operation=download.ListHandler ops="[download.ListHandler pool.List protocol.List vcsLister.List]" version=
time="2019-09-07T00:21:57Z" level=error msg="exit status 1: go list -m github.com: unrecognized import path \"github.com\" (parse https://github.com?go-get=1: no go-import meta tags ())\n" http-method=GET http-path=/github.com/@v/list http-url=/github.com/@v/list kind="Internal Server Error" module= operation=download.ListHandler ops="[download.ListHandler pool.List protocol.List vcsLister.List]" version=
time="2019-09-07T00:22:00Z" level=error msg="exit status 1: go list -m github.com/gorilla: invalid github.com/ import path \"github.com/gorilla\"\n" http-method=GET http-path=/github.com/gorilla/@v/list http-url=/github.com/gorilla/@v/list kind="Internal Server Error" module= operation=download.ListHandler ops="[download.ListHandler pool.List protocol.List vcsLister.List]" version=
time="2019-09-07T00:22:00Z" level=error msg="exit status 1: go list -m github.com: unrecognized import path \"github.com\" (parse https://github.com?go-get=1: no go-import meta tags ())\n" http-method=GET http-path=/github.com/@v/list http-url=/github.com/@v/list kind="Internal Server Error" module= operation=download.ListHandler ops="[download.ListHandler pool.List protocol.List vcsLister.List]" version=
time="2019-09-07T00:23:46Z" level=error msg="exit status 1: go list -m github.com/gorilla: invalid github.com/ import path \"github.com/gorilla\"\n" http-method=GET http-path=/github.com/gorilla/@v/list http-url=/github.com/gorilla/@v/list kind="Internal Server Error" module= operation=download.ListHandler ops="[download.ListHandler pool.List protocol.List vcsLister.List]" version=
time="2019-09-07T00:23:46Z" level=error msg="exit status 1: go list -m github.com: unrecognized import path \"github.com\" (parse https://github.com?go-get=1: no go-import meta tags ())\n" http-method=GET http-path=/github.com/@v/list http-url=/github.com/@v/list kind="Internal Server Error" module= operation=download.ListHandler ops="[download.ListHandler pool.List protocol.List vcsLister.List]" version=
time="2019-09-07T00:40:15Z" level=error msg="exit status 1: go list -m github.com/gorilla: invalid github.com/ import path \"github.com/gorilla\"\n" http-method=GET http-path=/github.com/gorilla/@v/list http-url=/github.com/gorilla/@v/list kind="Internal Server Error" module= operation=download.ListHandler ops="[download.ListHandler pool.List protocol.List vcsLister.List]" version=
time="2019-09-07T00:40:16Z" level=error msg="exit status 1: go list -m github.com: unrecognized import path \"github.com\" (parse https://github.com?go-get=1: no go-import meta tags ())\n" http-method=GET http-path=/github.com/@v/list http-url=/github.com/@v/list kind="Internal Server Error" module= operation=download.ListHandler ops="[download.ListHandler pool.List protocol.List vcsLister.List]" version=
time="2019-09-07T00:40:17Z" level=error msg="exit status 1: go list -m github.com/gorilla: invalid github.com/ import path \"github.com/gorilla\"\n" http-method=GET http-path=/github.com/gorilla/@v/list http-url=/github.com/gorilla/@v/list kind="Internal Server Error" module= operation=download.ListHandler ops="[download.ListHandler pool.List protocol.List vcsLister.List]" version=
time="2019-09-07T00:40:17Z" level=error msg="exit status 1: go list -m github.com: unrecognized import path \"github.com\" (parse https://github.com?go-get=1: no go-import meta tags ())\n" http-method=GET http-path=/github.com/@v/list http-url=/github.com/@v/list kind="Internal Server Error" module= operation=download.ListHandler ops="[download.ListHandler pool.List protocol.List vcsLister.List]" version=
time="2019-09-07T00:40:21Z" level=error msg="exit status 1: go list -m github.com/gorilla: invalid github.com/ import path \"github.com/gorilla\"\n" http-method=GET http-path=/github.com/gorilla/@v/list http-url=/github.com/gorilla/@v/list kind="Internal Server Error" module= operation=download.ListHandler ops="[download.ListHandler pool.List protocol.List vcsLister.List]" version=
time="2019-09-07T00:40:21Z" level=error msg="exit status 1: go list -m github.com: unrecognized import path \"github.com\" (parse https://github.com?go-get=1: no go-import meta tags ())\n" http-method=GET http-path=/github.com/@v/list http-url=/github.com/@v/list kind="Internal Server Error" module= operation=download.ListHandler ops="[download.ListHandler pool.List protocol.List vcsLister.List]" version=
time="2019-09-07T00:42:56Z" level=error msg="exit status 1: go list -m github.com/gorilla: invalid github.com/ import path \"github.com/gorilla\"\n" http-method=GET http-path=/github.com/gorilla/@v/list http-url=/github.com/gorilla/@v/list kind="Internal Server Error" module= operation=download.ListHandler ops="[download.ListHandler pool.List protocol.List vcsLister.List]" version=
time="2019-09-07T00:42:56Z" level=error msg="exit status 1: go list -m github.com: unrecognized import path \"github.com\" (parse https://github.com?go-get=1: no go-import meta tags ())\n" http-method=GET http-path=/github.com/@v/list http-url=/github.com/@v/list kind="Internal Server Error" module= operation=download.ListHandler ops="[download.ListHandler pool.List protocol.List vcsLister.List]" version=
```

**Comment by Skarlso** [comment]
I had this same error when trying to use Gaia which was downloading and building outside of project root. I'm trying to gather more information about this.

**Comment by marwan-at-work** [comment]
@marct83 TL;DR Athens version v0.5.0 uses Go 1.12 which has incompatible behavior than Go 1.13 

The Canary version (docker `gomods/athens:canary`) uses Go v1.13 as well as the correct behavior in returning 410 on incorrect import paths so that Go 1.13 doesn't fail. 

I also just released Athens v0.6.0. So please feel free to use that to pin to a stable version. 

I've run the test above using the following command: `docker run --rm -p 3000:3000 gomods/athens:v0.6.0`

And from a separate terminal, I successfully ran the following: 

```
$ GOPATH=$(tempdir) GO111MODULE=on GOPROXY='http://localhost:3000' go get -u github.com/gorilla/mux
go: finding github.com/gorilla/mux v1.7.3
go: downloading github.com/gorilla/mux v1.7.3
go: extracting github.com/gorilla/mux v1.7.3
```

Can you please confirm the behavior is fixed in v0.6.0 or canary? I'll close the issue for now, but feel free to open again. Thanks!

### More info for the curious: 

The reason you are seeing this is a change in Go 1.13's behavior is because Go 1.12 synchronously tried trimming the import path until it found the right module. However, Go 1.13 tries all trimmed import paths concurrently but it will fail if the proxy returned 500. Athens has always returned 500 for any `go mod download` and `go list -m` errors, because Go does not support proper error codes returned from that command.  However, we have changed this behavior to always return 404. In the future, I hope Go returns proper error codes so we distinguish expected and unexpected errors.
